### PR TITLE
Neuzz integration

### PR DESCRIPTION
**Pull request by ImanHosseini**
Build succeeds for all benchmarks except for arrow.

### CODE REVIEW

**Comment by ImanHosseini** [comment]
@googlebot I signed it!

File: fuzzers/neuzz/fuzzer.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+from fuzzers import utils
+
+WARMUP = 60 * 60
+
+
+def prepare_build_environment():
+    """Set environment variables used to build targets for AFL-based
+    fuzzers."""
+    cflags = ['-O2', '-fno-omit-frame-pointer']
+    # ASAN?
```
Comment (inferno-chromium) [inline fuzzers/neuzz/fuzzer.py:31]: This is problematic, can we not use existing flags. Setting your custom flags would mean it wont work with bug based benchmarks (they use ASan). It is better to remove/change part of flags that dont work (e.g. optimization flag).

**Review by inferno-chromium — CHANGES_REQUESTED** [review]

**Comment by inferno-chromium** [comment]
Thank you so much for bringing in Neuzz, we are very excited!

**Comment by lszekeres** [comment]
Thanks so much, Iman! 

Are there special requirements to run this? Do we need GPUs?

File: fuzzers/neuzz/fuzzer.py (hunk 1 above)
Comment (ImanHosseini) [reply fuzzers/neuzz/fuzzer.py]: done,

**Comment by inferno-chromium** [comment]
Need to run `make format` to clean up lint issues. Can verify everything is fine with `make presubmit`

File: fuzzers/neuzz/fuzzer.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    print(f'[run_neuzz] Ignoring additional flags: {additional_flags}')
+    command += ['--', target_binary]
+    print('[run_neuzz] Running command: ' + ' '.join(command))
+    output_stream = subprocess.DEVNULL if hide_output else None
+
+    proc = subprocess.Popen(command,
+                            stdout=subprocess.DEVNULL,
+                            stderr=subprocess.DEVNULL)
+
+    # subprocess.check_call(command, stdout=output_stream, stderr=output_stream)
```
[the remark this answers is not in the document]
Comment (ImanHosseini) [reply fuzzers/neuzz/fuzzer.py]: done. (except the line for custom flags, can we keep that? It's useful to comment/uncomment that to test)

File: fuzzers/neuzz/fuzzer.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+
+def run_neuzz(input_corpus,
+              output_corpus,
+              target_binary,
+              additional_flags=None,
+              hide_output=False):
+    """Run afl-fuzz."""
+    # Spawn the afl fuzzing process.
+    print('[run_neuzz] Running target with afl-fuzz')
```
[the remark this answers is not in the document]
Comment (ImanHosseini) [reply fuzzers/neuzz/fuzzer.py]: done.

File: fuzzers/neuzz/fuzzer.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    # Neuzz also requires afl-showmap.
+    print('[post_build] Copying afl-showmap to $OUT directory')
+    shutil.copy('/afl/afl-showmap', output_directory)
+    # Copy the Neuzz fuzzer itself.
+    print('[post_build] Copy neuzz fuzzer.')
+    shutil.copy('/neuzz/neuzz', output_directory)
+    shutil.copy('/neuzz/nn.py', output_directory)
+
+
+def prepare_fuzz_environment(input_corpus):
```
[the remark this answers is not in the document]
Comment (ImanHosseini) [reply fuzzers/neuzz/fuzzer.py]: done.

File: fuzzers/neuzz/runner.Dockerfile [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+FROM gcr.io/fuzzbench/base-image
+
+
+RUN apt-get update && \
+    apt-get install wget -y && \
+    apt-get install python-pip -y && \
+    python --version && \
+    python -m pip install --upgrade pip==20.3 && \
+    python -m pip install tensorflow==1.8.0 && \
+    python -m pip install keras==2.2.3
```
[the remark this answers is not in the document]
Comment (ImanHosseini) [reply fuzzers/neuzz/runner.Dockerfile]: done.

File: fuzzers/neuzz/fuzzer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (inferno-chromium) [reply fuzzers/neuzz/fuzzer.py]: yes can keep comments.

**Comment by ImanHosseini** [comment]
@inferno-chromium Sorry for the lint issues, I ran make format and it printed 'Success' but it wasn't fixing anything, I fixed the 'presubmit' lint errors manually in the end. Now presubmit succeeds (at least that's what I see) but spews 6 warnings which are regarding other code not related to neuzz.

@lszekeres as far as I know, since tensorflow 1.15 tensorflow-gpu is not separate, so if this would run with a gpu it would use it, otherwise fallback to run on cpu. Whether how that would impact the performance, I'm not really sure. Basically I took the old neuzz integration and updated it to work. I also haven't ran a day-long experiment for example, to see whether it crashes or something, but based on the original descriptions in the neuzz paper and repo, this should be the way to go.

File: fuzzers/neuzz/fuzzer.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+    # subprocess.check_call(command, stdout=output_stream, stderr=output_stream)
+    time.sleep(WARMUP)
+    proc.send_signal(signal.SIGINT)
+
+    print("[run_neuzz] Warmed up!")
+    command = [
+        "cp", "-RT", f"{output_corpus}/queue/", f"{input_corpus}_neuzzin/"
+    ]
+    print('[run_neuzz] Running command: ' + ' '.join(command))
```
[the remark this answers is not in the document]
Comment (ImanHosseini) [reply fuzzers/neuzz/fuzzer.py:80]: done.

**Review by ImanHosseini — COMMENTED** [review]
Applied the changes.

**Review by inferno-chromium — APPROVED** [review]
AMAZING, looks like CI worked.

**Pull request merged**